File: cpp/ops/Math/math.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include "math.h"
#include "util.h"
#include <iostream>
#include <cmath>

namespace math {

using namespace std;

void plus(float *a, float *b, float *res, int size) {
  for (size_t i = 0; i < size; i++)
    res[i] = a[i] + b[i];
}

void minus(float *a, float *b, float *res, int size) {
  for (size_t i = 0; i < size; i++)
    res[i] = a[i] - b[i];
}

void unaryMinus(float *a, float *res, int size) {
  for (size_t i = 0; i < size; i++)
    res[i] = -a[i];
}

void times(float *a, float *b, float *res, int size) {
  for (size_t i = 0; i < size; i++)
    res[i] = a[i] * b[i];
}

void exp(float *a, float *res, int size) {
  for (size_t i = 0; i < size; i++)
    res[i] = std::exp(a[i]);
}

void log(float *a, float *res, int size) {
  for (size_t i = 0; i < size; i++)
    res[i] = std::log(a[i]);
}

void lgamma(float *a, float *res, int size) {
  for (size_t i = 0; i < size; i++)
    res[i] = std::lgamma(a[i]);
}

void digamma(float *a, float *res, int size) {
  for (size_t i = 0; i < size; i++)
    res[i] = util::digamma(a[i]);
}

void trigamma(float *a, float *res, int size) {
  for (size_t i = 0; i < size; i++)
    res[i] = util::trigamma(a[i]);
}

void polygamma(int n, float *a, float *res, int size) {
  if (n == 0) {
    digamma(a, res, size);
  } else if (n == 1) {
    trigamma(a, res, size);
  } else {
    for (size_t i = 0; i < size; i ++)
      res[i] = util::polygamma(n, a[i]);
  }
}

float lgamma(float f) {
  return std::lgamma(f);
}

float digamma(float f) {
  return util::digamma(f);
}

float polygamma(int n, float f) {
  if (n == 0) {
    return util::digamma(f);
  } else if (n == 1) {
    return util::trigamma(f);
  } else {
    return util::polygamma(n, f);
  }
}

} // namespace math

File: kotlin/api/src/main/kotlin/org/diffkt/external/Math.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package org.diffkt.external

internal object Math {

    fun plus(a: FloatArray, b: FloatArray, size: Int): FloatArray {
        val res = FloatArray(size)
        External.plus(a, b, res, size)
        return res
    }

    fun minus(a: FloatArray, b: FloatArray, size: Int): FloatArray {
        val res = FloatArray(size)
        External.minus(a, b, res, size)
        return res
    }

    fun unaryMinus(a: FloatArray, size: Int): FloatArray {
        val res = FloatArray(size)
        External.unaryMinus(a, res, size)
        return res
    }

    fun exp(a: FloatArray, size: Int): FloatArray {
        val res = FloatArray(size)
        External.exp(a, res, size)
        return res
    }

    fun log(a: FloatArray, size: Int): FloatArray {
        val res = FloatArray(size)
        External.log(a, res, size)
        return res
    }

    fun times(a: FloatArray, b: FloatArray, size: Int): FloatArray {
        val res = FloatArray(size)
        External.times(a, b, res, size)
        return res
    }

    fun lgamma(a: FloatArray): FloatArray {
        val res = FloatArray(a.size)
        External.lgamma(a, res, a.size)
        return res
    }

    fun digamma(a: FloatArray): FloatArray {
        val res = FloatArray(a.size)
        External.digamma(a, res, a.size)
        return res
    }

    fun polygamma(n: Int, a: FloatArray): FloatArray {
        val res = FloatArray(a.size)
        External.polygamma(n, a, res, a.size)
        return res
    }

    // Scalars going to C++ for now.
    fun lgamma(f: Float): Float {
        return External.lgamma(f)
    }

    fun digamma(f: Float): Float {
        return External.digamma(f)
    }

    fun polygamma(n: Int, f: Float): Float {
        return External.polygamma(n, f)
    }

}
